gflownet/envs/crystals/spacegroup.py

```diff
@@ -539,8 +559,61 @@ def get_ref_index(self, state: List[int] = None) -> int:
             state = self.state
         return sum([int(s > 0) * f for s, f in zip(state, self.ref_state_factors)])
 
-    def get_all_terminating_states(self) -> List[List]:
+    def set_n_atoms(self, composition):
+        if composition is None:
+            self.n_atoms = None
+        else:
+            self.n_atoms = [n for n in composition if n > 0]
+        # Get compatibility with stoichiometry
+        self.compatibility_stoichiometry_dict = (
+            SpaceGroup.get_compatibility_stoichiometry_dict(
+                self.n_atoms, self.space_groups.keys()
+            )
+        )
+        return self.n_atoms, self.compatibility_stoichiometry_dict
+
+    @staticmethod
+    def get_compatibility_stoichiometry_dict(
+        n_atoms: List[int], space_groups: List[int]
+    ):
+        """
+        Obtains which space groups are compatible with the stoichiometry given as
+        argument (n_atoms). It relies on pyxtal's
+        pyxtal.symmetry.Group.check_compatible(). Note that True is stored only if both
+        is_compatible and has_freedom are True.
```

Comment:
To be honest, I do not really understand the meaning of `has_freedom`. I had interpreted it as something that should be True. Are you confident that it can be False?

Reply:
I did some more testing in the same vein as our discussion on Slack and it seems to support how I understood this method.

For space group 227 (https://www.cryst.ehu.es/cgi-bin/cryst/programs/nph-wp-list) : 
- the smallest wyckoff positions (8a, 8b) have a multiplicity of 8 but have no degree of freedom; if we "pick" those positions, we don't get to choose an x, y, or z.
- there are also positions (16c, 16d) with multiplicity of 16 but they are also fixed
- the most specific wyckoff position where we can decide a x, y or z is 32e. This position has a degree of freedom.

If I try with a composition incompatible with the space group, I get `has_freedom=False`. 
If I try with various compositions involving the positions with multiplicity 8 and 16, I get `has_freedom=False`.
As soon as I go high enough in the number of atoms to allow position 32e to come in the picture, I get `has_freedom=True`. It stays true even if I give compositions that will force it to combine 32e with a fixed wyckoff position like 8a or 8b.

```
>>> from pyxtal.symmetry import Group
>>> Group(227).check_compatible([2])
(False, False)
>>> Group(227).check_compatible([8])
(True, False)
>>> Group(227).check_compatible([16])
(True, False)
>>> Group(227).check_compatible([32])
(True, True)
>>> Group(227).check_compatible([40])
(True, True)
>>> Group(227).check_compatible([40, 8])
(True, True)
```

At the ends of the day, my interpretation of `has_freedom` is that is means that there exists at least one solution to the WP assignment problem in which at least one of the WP used doesn't have fixed X, Y and Z. Based on that, I don't see a reason to refuse the cases where `has_freedom` would be false.

And, actually, I see a reason why we should accept the cases in which `has_freedom` is False (as long as `is_valid` is True). And that reason is that this : 
```
>>> Group(227).check_compatible([8])
(True, False)
```
Corresponds to a real, well studied crystal : https://next-gen.materialsproject.org/materials/mp-149/
If we required that has_freedom be True, we would be preventing the GFN from generating stuff that we know is valid.

Reply:
This is a great explanation and analysis. Thanks for looking into this! I am going to change the code to check only the first returned element `is_valid`.

Reply:
Sadly, it doesn't imply that there is only one possible arrangements. Only that at least one of the possible arrangements used non-fixed WP.

For instance, for the composition [8, 8], it returns : 
```
>>> from pyxtal.symmetry import Group
>>> Group(227).check_compatible([8, 8])
(True, False)
```
In this case, there are two valid arrangements of atoms. Either the first 8 atoms get assigned to WP 8a and the next 8 atoms get assigned to 8b, or the reverse order. But all of these possible arrangements involve only Wyckoff positions with fixed coordinates so `has_freedom` is False even if there are multiple arrangements.

Reply:
Oh I see! But it _does_ imply that there is a **finite** number of arrangements, is that correct? That would also be valuable info.

Reply:
I think that is correct.